File: python/util/primal.py
import math


# primes using Sieve of Eratosthenes (storing only odds)
def get_primes(limit):
  odds_only_limit = limit // 2
  primes = [True] * odds_only_limit
  primes[0] = False

  for i in range(1, math.floor(math.sqrt(limit) + 1)):
    n = 2 * i + 1
    if primes[i]:
      step = n
      start = i + step if step == 3 else i + step * 2
      for j in range(start, odds_only_limit, step):
        primes[j] = False

  return primes


# get prime numbers by passing in the sieve
def get_prime_numbers(primes):
  prime_nums = [2]
  for i, prime_bool in enumerate(primes):
    if prime_bool:
      prime_nums.append(2 * i + 1)
  return prime_nums


def is_prime(num, primes=None):
  if primes is not None and len(primes) >= num:
    if num % 2 == 0:
      return num == 2
    return primes[math.floor((num - 1) / 2)]

  # primes (sieve) wasn't passed as an argument, naive check if n is prime
  if num < 1:
    return True
  elif num == 2 or num == 3:
    return True
  elif num % 2 == 0 or num % 3 == 0:
    return False
  i = 5
  while i * i <= num:
    if num % i == 0 or num % (i + 2) == 0:
      return False
    i += 6

  return True


def get_prime_factors(num, primes):
  if is_prime(num, primes):
    return []

  prime_factors = []
  if num % 2 == 0:
    base_exp = {'base': 2, 'exp': 0}
    while num % 2 == 0:
      base_exp['exp'] += 1
      num /= 2
    prime_factors.append(base_exp)

  sqrt = math.sqrt(num)
  for i in range(3, math.floor(sqrt + 2), 2):
    if num < i:
      break
    if is_prime(i, primes):
      if num % i == 0:
        base_exp = {'base': i, 'exp': 0}
        while num % i == 0:
          base_exp['exp'] += 1
          num /= i
        prime_factors.append(base_exp)

  if num > 2:
    prime_factors.append({'base': num, 'exp': 1})

  return prime_factors


# sieve factor counts
def get_prime_factor_counts(limit):
  prime_factor_counts = [0] * (limit + 1)
  prime_factor_counts[1] = 1

  for i in range(2, len(prime_factor_counts)):
    if prime_factor_counts[i] != 0:
      continue

    step = i
    for j in range(i, len(prime_factor_counts), step):
      prime_factor_counts[j] += 1

  return prime_factor_counts

File: javascript/util/primal.js
// primes using Sieve of Eratosthenes (storing only odds)
function getPrimes(limit) {
  const oddsLimit = Math.floor(limit / 2);
  const primes = Array(oddsLimit).fill(true);

  primes[0] = false;

  for (let i = 1; i <= Math.sqrt(limit); i++) {
    const n = 2 * i + 1;
    if (primes[i]) {
      const step = n;
      for (let j = step === 3 ? i + step : i + step * 2; j <= oddsLimit; j += step) {
        primes[j] = false;
      }
    }
  }

  return primes;
}

// get prime numbers by passing in the sieve
function getPrimeNumbers(primes) {
  return primes.reduce(
    (a, c, i) => {
      if (c) {
        a.push(2 * i + 1);
      }
      return a;
    },
    [2],
  );
}

function isPrime(num, primes) {
  if (primes && primes.length && primes.length >= num) {
    if (num % 2 === 0) {
      return num === 2;
    }

    return primes[(num - 1) / 2];
  }

  // primes (sieve) wasn't passed as an argument, naive check if n is prime
  if (num < 1) {
    return true;
  }
  if (num === 2 || num === 3) {
    return true;
  }
  if (num % 2 === 0 || num % 3 === 0) {
    return false;
  }

  let i = 5;
  while (i * i <= num) {
    if (num % i === 0 || num % (i + 2) === 0) {
      return false;
    }
    i += 6;
  }

  return true;
}

function getPrimeFactors(num, primes) {
  if (isPrime(num, primes)) {
    return [];
  }

  const primeFactors = [];
  if (num % 2 === 0) {
    const baseExp = {
      base: 2,
      exp: 0,
    };
    while (num % 2 === 0) {
      baseExp.exp++;
      num /= 2;
    }
    primeFactors.push(baseExp);
  }

  const sqrt = Math.sqrt(num);
  for (let i = 3; i <= sqrt; i += 2) {
    if (num < i) {
      break;
    }
    if (isPrime(i, primes)) {
      if (num % i === 0) {
        const baseExp = {
          base: i,
          exp: 0,
        };
        while (num % i === 0) {
          baseExp.exp++;
          num /= i;
        }
        primeFactors.push(baseExp);
      }
    }
  }

  if (num > 2) {
    primeFactors.push({
      base: num,
      exp: 1,
    });
  }

  return primeFactors;
}

// sieve factor counts
function getPrimeFactorCounts(limit) {
  const primeFactorCounts = Array(limit + 1).fill(0);
  primeFactorCounts[1] = 1;

  for (let i = 2; i < primeFactorCounts.length; i++) {
    if (primeFactorCounts[i] !== 0) {
      continue;
    }

    const step = i;
    for (let j = i; j < primeFactorCounts.length; j += step) {
      primeFactorCounts[j]++;
    }
  }

  return primeFactorCounts;
}

module.exports = {
  getPrimes,
  getPrimeNumbers,
  isPrime,
  getPrimeFactors,
  getPrimeFactorCounts,
};
